### PR TITLE
Channel Bookmarks

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This introduces a feature called channel bookmarks that will allow users to publish links or files to the top of a channel for quick access.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-54416

#### Release Note

```release-note
Added Channel bookmarks
```

### CODE REVIEW

**Comment by lindy65** [comment]
Hey @enahum - please add a link to the Jira ticket in the description. Thanks!

File: server/channels/db/migrations/postgres/000120_create_channelbookmarks_table.up.sql
Comment (agnivade) [inline server/channels/db/migrations/postgres/000120_create_channelbookmarks_table.up.sql:14]: You can create an enum type as `CREATE TYPE bookmark_type AS ENUM ('link', 'file');` and then use `bookmark_type` as the column type. See `create_enums.up.sql`

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:88]: Please create a custom type and use sqlx.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:115]: This needs to be done in the same transaction.

Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:115]: This needs to be done in the same transaction.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:181]: Avoid using `sql.ErrNoRows` directly in the code. It's something that the underlying sql driver will return. From the application, use `store.NewErrNotFound` for anything that's not found. Handle this in the application side, and construct an app error accordingly.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:225]: Some things to be fixed here:
1. Number of bookmarks seem to be unbounded here. We need to fix that and enforce a limit from the application side.
2. We need to execute a single SQL statement, rather than a series of UPDATEs.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:296]: Same comment as earlier. Use a custom struct and use sqlx.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:264]: Please use squirrel to avoid SQL injections.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:275]: There is no `LIMIT` in the SQL query. We need that to avoid performance and security issues.

**Review by agnivade — CHANGES_REQUESTED** [review]

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (enahum) [reply server/channels/store/sqlstore/channel_bookmark_store.go:225]: > We need to execute a single SQL statement, rather than a series of UPDATEs.

How do you propose we should do this in a single statement?

Comment (enahum) [reply server/channels/store/sqlstore/channel_bookmark_store.go]: ended up using a case in a single update query

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:278]: Sorry for not being clear. I didn't mean to suggest a hardcoded limit. Rather, we need to set this limit from the application side. Does the application throw an error if a user tries to add a 1001th bookmark? If not, then this would be a bug to simply cut off at 1000th entry.

Either we paginate based on batches and iterate them from the client side, or we simply set a limit from the UI. If we decide to do the latter, then the limit needs to come from the app layer. The store layer should not have any hardcoding.

**Review by agnivade — COMMENTED** [review]
Great work Elias! Just one final comment from my side.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (enahum) [reply server/channels/store/sqlstore/channel_bookmark_store.go:278]: Ok I get what you are saying here.

Is there a problem if we just query them in batches but include all of them in the response?

File: server/channels/app/channel_bookmark.go
Comment (agnivade) [inline server/channels/app/channel_bookmark.go:39]: `c` should be used to access the logger: `c.Logger()`

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (agnivade) [inline server/channels/store/sqlstore/channel_bookmark_store.go:93]: `COUNT(*) as count` for better performance. Refer: https://docs.google.com/presentation/d/14LTiD4EHrlCgVfMFngnWe1Ov6gsbgwXMGp3b6jM5p3A/edit#slide=id.g1092ca99953_0_87

**Review by agnivade — APPROVED** [review]
Just 2 small comments. Good to go after those are handled.

FYI - I have mainly focused on DB and performance aspect. Will leave it to someone else to verify the business logic.

**Comment by jupenur** [comment]
@enahum @esarafianou how much of a priority is this? 9.6 feature complete is in two weeks, is that what we're aiming for?

**Comment by enahum** [comment]
> @enahum @esarafianou how much of a priority is this? 9.6 feature complete is in two weeks, is that what we're aiming for?

Yes the idea is to merge asap, although webapp work is still on progress.

Perhaps @esethna can share more about the target date

**Comment by esethna** [comment]
@jupenur @esarafianou as Elias mentioned we're trying to merge this asap so the team can move onto other projects. Are there blockers?

**Comment by agnivade** [comment]
I would urge not to rush anything here and let the security review happen. Being "reliable and stable" is also an objective that we want to pursue, and it's important that we deliver a feature that works at scale and under all conditions.

File: server/channels/app/channel_bookmark.go
Comment (isacikgoz) [inline server/channels/app/channel_bookmark.go:39]: What was the reason of adding additional log line here? IIRC it should be printed if the error is =< 500 anyway in the API layer.

File: server/channels/app/channel_bookmark_test.go
Comment (isacikgoz) [inline server/channels/app/channel_bookmark_test.go:47]: We shouldn't use `assert` here as `bookmarkResp` can be `nil` if there is an error. It would cause panics. There are similar occurrences throughout the file, can you double check?

**Review by isacikgoz — CHANGES_REQUESTED** [review]
This is a very neat work, I have a few comments initially but overall this looks good. Some changes may required to avoid panics in the tests.

**Comment by enahum** [comment]
> This is a very neat work, I have a few comments initially but overall this looks good. Some changes may required to avoid panics in the tests.

@isacikgoz Thanks for the review, I will address the comments on Monday :)

File: server/channels/api4/apitestlib.go
Comment (enahum) [reply server/channels/api4/apitestlib.go:637]: is there anything for me to address here?

Comment (isacikgoz) [reply server/channels/api4/apitestlib.go:637]: No it was just a wish :)

**Review by isacikgoz — APPROVED** [review]
Awesome @enahum, looks even better :) Thanks!!

**Comment by agnivade** [comment]
@esarafianou / @jupenur - Just to provide some context here from my side. When I was reviewing the PR, I found some issues with unbounded payloads and pagination, which has been now addressed. That's why I asked for a security review in case there were other hidden issues that escaped me.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (jupenur) [inline server/channels/store/sqlstore/channel_bookmark_store.go:146]: What if this file is already attached to a post or another channel bookmark?

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (jupenur) [reply server/channels/store/sqlstore/channel_bookmark_store.go:283]: Agniva is right, this needs a limit.

File: server/channels/app/channel_bookmark.go
Comment (jupenur) [inline server/channels/app/channel_bookmark.go:63]: What if this file is already attached somewhere else?

File: server/channels/app/channel_bookmark.go
Comment (jupenur) [inline server/channels/app/channel_bookmark.go:83]: Here, too: what if the FileId on this bookmark is attached somewhere else?

**Review by jupenur — CHANGES_REQUESTED** [review]
Some quick initial comments. Review in progress.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (enahum) [reply server/channels/store/sqlstore/channel_bookmark_store.go:283]: This was addressed by adding a cap of 50 items per channel, I can still add the limit here if we want to

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (enahum) [reply server/channels/store/sqlstore/channel_bookmark_store.go:146]: This is a good question that I guess it also applies to posts. I ron't think we have any rule on this, do we want to enforce one?

Comment (jupenur) [reply server/channels/store/sqlstore/channel_bookmark_store.go:146]: Posts handle this here: https://github.com/mattermost/mattermost/blob/8395575d11797583174129c1220051842870d175/server/channels/store/sqlstore/file_info_store.go#L397-L404

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (jupenur) [reply server/channels/store/sqlstore/channel_bookmark_store.go:283]: This method returns deleted bookmarks, the cap only applies to non-deleted ones.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (enahum) [reply server/channels/store/sqlstore/channel_bookmark_store.go:186]: done

File: server/channels/app/channel_bookmark.go
Comment (enahum) [reply server/channels/app/channel_bookmark.go:71]: done

File: server/channels/app/channel_bookmark.go
Comment (jupenur) [inline server/channels/app/channel_bookmark.go:54]: What if the bookmark didn't have a file attached before? Wouldn't this `isAnotherFile` then be false and the new file wouldn't go through the necessary checks?

**Review by jupenur — CHANGES_REQUESTED** [review]
Slowly working my way through, review still very much in progress.

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (calebroseland) [reply server/channels/store/sqlstore/channel_bookmark_store.go:289]: +1 https://mattermost.atlassian.net/browse/MM-56774

File: server/channels/app/channel_bookmark.go
Comment (enahum) [reply server/channels/app/channel_bookmark.go:54]: The model is not valid in that scenario.

All bookmarks of type link must have a url, all bookmarks of type file must have a file

Comment (jupenur) [reply server/channels/app/channel_bookmark.go:54]: But a bookmark of type link can also have a file, at least as far as I can tell

File: server/channels/store/sqlstore/channel_bookmark_store.go
Comment (enahum) [reply server/channels/store/sqlstore/channel_bookmark_store.go]: done

File: server/channels/app/channel_bookmark.go
Comment (enahum) [reply server/channels/app/channel_bookmark.go:54]: done

Comment (jupenur) [reply server/channels/app/channel_bookmark.go:54]: Looks like the Patch method can also change the bookmark type?

File: server/channels/store/sqlstore/file_info_store.go
Comment (enahum) [reply server/channels/store/sqlstore/file_info_store.go:473]: Yes this was included in the spec!

And you bring a good point, we can make sure the attachment belongs to a bookmark.

As for the deleted file, it should not crash but the bookmark won't open

**Review by jupenur — APPROVED** [review]
Looks like you could upload a file in one channel and then attach it as a bookmark in another, and then you'd have a file bookmark with the bookmark object pointing to one channel and the fileinfo object pointing to another, and it probably wouldn't work very well.

That's the last thing I spotted though, and I don't think it has any security impact, so approving from the security PoV.

**Comment by jupenur** [comment]
The `createChannelBookmark` endpoint allows specifying a user-defined ID for your new bookmark and `PreSave` only generates a new ID if there isn't one yet. That doesn't seem intentional?

**Comment by enahum** [comment]
> The `createChannelBookmark` endpoint allows specifying a user-defined ID for your new bookmark and `PreSave` only generates a new ID if there isn't one yet. That doesn't seem intentional?

addressed.

**Pull request merged**
